### PR TITLE
REFACTOR: Move all section related code to action creators and store.

### PR DESCRIPTION
**Pull request by thepug**
Actions involved:

```
:boards-load-other
:board-get
:board-seen
:board-nav-away
:board
:board-delete
:whats-new/finish
:section-edit-save
:section-edit-save/finish
:section-edit/dismiss
:private-section-user-add
:private-section-user-remove
:private-section-kick-out-self
:private-section-kick-out-self/finish
:container/section-change
```

To test:

`:boards-load-other`
- Login and go to all posts.
- [x] In the console do you see requests for the existing sections? `/boards/<board-name>` followed by `Dispatching action: :section`

`:board-get`  This also tests `:board` (the old way of handling the API return)
Technically was tested above but you can also
- Login and click on a section.
- [x] The posts for that section load and you will see the same `Dispatching action :section` message.
Draft test
- Create a new draft
- [x] Does the draft section load with the draft post?

`:board-seen`
Repeating the same steps for `:board-get` 
- [x] Do you see `Dispatching action :section-seen` in the console?

`:board-nav-away`
- Switch boards from the `:board-get` test. You have to be in the old "entries" view .
- [x] Do you see `Dispatching action :section-nav-away`?

`:board-delete`
- [x] Is the section deleted when clicking on the section editor and clicking delete?

```
:section-edit-save
:section-edit-save/finish
:section-edit/dismiss
```
- [x] Does create a new section work?
- [x]  Does editing the section name work?

```
:private-section-user-add
:private-section-user-remove
:private-section-kick-out-self
:private-section-kick-out-self/finish
```
- [x] Make a board private and add yourself. Does saving this work?
- [x]  Make a board private and add someone else.
- [x] Remove the other user and save.
- [x] Remove yourself and save.

`:container/section-change`
This was added to handle websocket messages from the change service.
- Login as user A and user B
- As user A click on a section.
- [x] As user B add a post to the same section and does it show up for user A?

### CODE REVIEW

File: src/oc/web/actions.cljs [hunk 1]
```diff
@@ -59,10 +60,10 @@
       (if-let [board-data (first (filter #(= (:slug %) (router/current-board-slug)) boards))]
         ; Load the board data since there is a link to the board in the org data
         (when-let [board-link (utils/link-for (:links board-data) ["item" "self"] "GET")]
-          (api/get-board board-link))
-        ; The board wasn't found, showing a 404 page
+          (sa/section-get board-link))
+
         (if (= (router/current-board-slug) utils/default-drafts-board-slug)
-          (utils/after 100 #(dispatcher/dispatch! [:board utils/default-drafts-board]))
+          (utils/after 100 #(sa/section-get utils/default-drafts-board))
```
Comment (bago2k4) [inline src/oc/web/actions.cljs]: @thepug this needs to dispatch the board data without loading it. It's needed to make sure that the user doesn't end up to a 404 when visiting the drafts board if it's empty.
The storage doesn't include a drafts board in the org boards list if you don't have any but can be that the user force the url to it, in this case i forge the draft board data (check utils/default-drafts-board)

**Comment by belucid** [comment]
@thepug conflict here now.

**Comment by thepug** [comment]
@belucid ok, conflicts resolved

**Comment by belucid** [comment]
Reviewing again.

**Comment by belucid** [comment]
@thepug how come when we nav. section to section we're refreshing all sections?

**Comment by thepug** [comment]
> @thepug how come when we nav. section to section we're refreshing all sections?

I really don't know. I think we use it to check for new sections and org data, so the request is to get the org data which returns a list of sections.  It does seem like we should only load this data when notified by the change service.  Also I think there are some issues with all posts that require getting all sections.

**Comment by belucid** [comment]
@bago2k4 do you know why we are loading all sections when nav'ing between 1 section and another?

**Comment by belucid** [comment]
@thepug the last test fails.

- User A on Section 1
- User B on Section 1
- User B adds a new post to Section 1
- User A gets the carrot spinner and is toast

Here's User A's console:

```
DEBUG [oc.web.lib.ws-change-client:102] - Push event from server:  [:container/change {:container-id "f213-4509-8774", :change-at "2018-04-02T12:53:54.326Z"}]
raven.inc.js:58:7
DEBUG [oc.web.lib.ws-change-client:55] - event-handler :container/change
raven.inc.js:58:7
DEBUG [oc.web.lib.ws-change-client:73] - Change event: {:container-id "f213-4509-8774", :change-at "2018-04-02T12:53:54.326Z"}
raven.inc.js:58:7
DEBUG [oc.web.actions.section:61] - Section change: f213-4509-8774 at: 2018-04-02T12:53:54.326Z
raven.inc.js:58:7
INFO [oc.web.dispatcher:317] - Dispatching action: :section-change
raven.inc.js:58:7
DEBUG [oc.web.stores.reaction:21] - Dispatching reaction reducer: :section-change
raven.inc.js:58:7
INFO [oc.web.dispatcher:317] - Dispatching action: :container/change
raven.inc.js:58:7
DEBUG [oc.web.actions:32] - Container change: f213-4509-8774 at: 2018-04-02T12:53:54.326Z by: nil
raven.inc.js:58:7
DEBUG [oc.web.stores.reaction:21] - Dispatching reaction reducer: :container/change
raven.inc.js:58:7
INFO [oc.web.dispatcher:317] - Dispatching action: :section-nav-away
raven.inc.js:58:7
DEBUG [oc.web.stores.section:48] - Section nav away: f213-4509-8774
raven.inc.js:58:7
DEBUG [oc.web.stores.reaction:21] - Dispatching reaction reducer: :section-nav-away
```

**Comment by thepug** [comment]
RE: https://github.com/open-company/open-company-web/pull/435#issuecomment-377923258

@belucid There were some conditionals not returning an app state which causes the loading carrot.
Addressed those.

**Comment by thepug** [comment]
[Step 1: Clean up actions in web app](https://trello.com/c/nOF8CduC/1102-step-1-clean-up-actions-in-web-app)

**Comment by bago2k4** [comment]
Yeah it's here:
https://github.com/open-company/open-company-web/blob/section-refactor/src/oc/web/components/org_dashboard.cljs#L51

When org-dashboard component remount it refresh the data. Now this is needed in case the user is moving to AP since we don't refresh AP data on change service messages, but it's not needed in case the user moves to a board (that should be already up to date due to change messages).

I think we should remove line 51 (don't refresh current board and subsequentially the other boards on remount) and also the `load-other-boards` at line 46 (link below).

https://github.com/open-company/open-company-web/blob/section-refactor/src/oc/web/components/org_dashboard.cljs#L46

This may have been my fault on an old PR, suggesting to Nathan to refresh those things here.

**Comment by bago2k4** [comment]
It actually may be a bit more complicated than what i suggested above.
But the point is that we need to make sure of these points:
- if the user is looking at a single board
  1) load the board the user is looking at
  2) preload all the other boards to have them ready to display
  3) preload the AP data?
- if the user is looking ad AP
  1) load AP data
  2) preload all the other boards

In all cases make sure we refresh the the changed board. What about AP? We need to refresh it when the user actually navigate to it? or on every change?

Before now it was working because we used to refresh the complete boards list on every user navigation.

**Pull request merged**